engine/engine_test.go

```diff
@@ -347,6 +347,242 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
+	gotPutAdCid1, err := subject.NotifyPut(ctx, "", []string{"/ip4/0.0.0.0/tcp/1234/http"}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+
+	gotLatestAdCid, ad, err := subject.GetLatestAdv(ctx)
+	require.NoError(t, err)
+	require.Equal(t, gotLatestAdCid, gotPutAdCid1)
+	require.Equal(t, ad.Provider, subject.ProviderID())
+	require.Equal(t, ad.Addresses, subject.ProviderAddrs())
+}
+
+func TestEngine_VerifyErrAlreadyAdvertised(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs := testutil.RandomMultihashes(t, rng, 42)
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, "", []string{}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+
+	_, err = subject.NotifyPut(ctx, "", []string{}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.Error(t, err, provider.ErrAlreadyAdvertised)
+
+	_, err = subject.NotifyPut(ctx, "SomeOtherProvider", []string{}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err, provider.ErrAlreadyAdvertised)
+}
+
+func TestEngine_PublishingOfTheSameCIDsForDifferentProviders_ShouldResultIntoSameChunksInChunker(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs := testutil.RandomMultihashes(t, rng, 42)
+
+	provider1id := "Provider 1"
+	provider1Addrs := []string{"/ip4/0.0.0.0/tcp/1234/http"}
+	provider2id := "Provider 2"
+	provider2Addrs := []string{"/ip4/0.0.0.0/tcp/4321/http"}
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, provider1id, provider1Addrs, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+	require.Equal(t, 1, subject.Chunker().Len())
+
+	gotPutAdCid2, err := subject.NotifyPut(ctx, provider2id, provider2Addrs, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid2)
+	require.Equal(t, 1, subject.Chunker().Len())
```

Comment:
Would it make sense to also assert that ad entries link is the same across both ads?

Reply:
Good point, added